File: 19/robots.ts
import { promises as fs } from 'fs';
import * as path from 'path';

const RESOURCE_TYPES = ['ore', 'clay', 'obsidian', 'geode'] as const;

const ZERO_RESOURCE_MAP = new Map(RESOURCE_TYPES.map((type) => [type, 0]));

class Action {
    type: string = '';
    costs: Map<string, number> = new Map();

    toString(): string {
        return `${this.type} ${this.costs}`;
    }
}

function mapToString(map: Map<string, number>): string {
    return Array.from(map.entries())
        .map(([key, value]) => `${key}=${value}`)
        .join(' ');
}

/**
 * Parses the cost of each robot type from a line of the input using regex.
 *
 * Example line:
 * `Blueprint 1: Each ore robot costs 4 ore. Each clay robot costs 4 ore. Each obsidian robot costs 4 ore and 17 clay. Each geode robot costs 4 ore and 20 obsidian.`
 */
function parseCostsToActions(line: string): Action[] {
    // First pull out the sentence that describes the costs of each robot type.
    const robotRegex = /Each (ore|clay|obsidian|geode) robot costs (.*?)\./g;
    const robotMatches = [...line.matchAll(robotRegex)];
    // Then parse the costs of each robot type.
    const actions: Action[] = [];
    for (const robotMatch of robotMatches) {
        const action = new Action();
        const robotType = robotMatch[1];
        action.type = robotType;
        const costs = robotMatch[2];
        const costRegex = /(\d+) (ore|clay|obsidian|geode)/g;
        const costMatches = [...costs.matchAll(costRegex)];
        for (const costMatch of costMatches) {
            const cost = parseInt(costMatch[1]);
            const resourceType = costMatch[2];
            action.costs.set(resourceType, cost);
        }
        actions.push(action);
    }
    return actions;
}

function maximiseGeodes(
    actions: Action[], timeLeft: number, resources: Map<string, number>,
    resourcesPerMinute: Map<string, number>, statesVisited: Map<string, number>, depth: number = 0): number {
    const depthStr = ' '.repeat(depth);
    // console.log(`${depthStr}maximiseGeodes(${timeLeft}, Resources: ${mapToString(resources)}, ResourcesPerMin: ${mapToString(resourcesPerMinute)})`);

    if (timeLeft < 0) {
        return 0;
    }

    const resourcesPerMinString = mapToString(resourcesPerMinute);

    // This might eliminate some valid solutions, but the idea is if we've had
    // the same amount of robots in a previous state, but this time we have less
    // time left, then this state must be worse than the previous state so we
    // can stop the search here.

    // Ok, is does eliminate some solutions. I guess lets just add some kludge factor here.
    const previousTimeLeft = statesVisited.get(resourcesPerMinString) ?? 0;
    if (previousTimeLeft > timeLeft + 1) {
        return 0;
    }
    else {
        statesVisited.set(resourcesPerMinString, Math.max(previousTimeLeft, timeLeft));
    }

    // How many geodes we'd have if we waited and did nothing else.
    let bestGeodes = resources.get('geode')! + timeLeft * resourcesPerMinute.get('geode')!;



    // For each type of robot, try waiting until we have enough resources and then building it.
    actionLoop: for (const action of actions) {
        // console.log(`${depthStr}Trying to build ${action.type} robot...`)
        // First wait until we have enough resources.
        let timeToWait = 0;
        const resourcesNeeded = new Map(action.costs);
        for (const [resource, amount] of resourcesNeeded) {
            const amountLeft = amount - resources.get(resource)!;
            const amountPerMinute = resourcesPerMinute.get(resource)!;
            // No amount of waiting will ever give us enough resources.
            if (amountLeft > 0 && amountPerMinute === 0) {
                // console.log(`${depthStr}Not enough ${resource} to build ${action.type} robot.`);
                continue actionLoop;
            }
            const timeToWaitForThisResource = Math.ceil(amountLeft / amountPerMinute);
            timeToWait = Math.max(timeToWait, timeToWaitForThisResource);
        }
        // If we don't have enough time left to wait, then we can't build this robot.
        if (timeToWait > timeLeft) {
            continue;
        }

        // We need to wait 1 extra minute to build the robot.
        timeToWait++;

        // Start another search after building this robot.
        const newResources = new Map();
        const newResourcesPerMinute = new Map();
        for (const resource of RESOURCE_TYPES) {
            const newAmount = resources.get(resource)! - (resourcesNeeded.get(resource) ?? 0) + timeToWait * resourcesPerMinute.get(resource)!;
            const newAmountPerMinute = resourcesPerMinute.get(resource)! + (resource === action.type ? 1 : 0);
            newResources.set(resource, newAmount);
            newResourcesPerMinute.set(resource, newAmountPerMinute);
        }

        // console.log(`${depthStr}Building ${action.type} robot after waiting ${timeToWait} minutes. ` +
            // `Resources: ${mapToString(newResources)}. Resources per minute: ${mapToString(newResourcesPerMinute)}.`);

        const geodes = maximiseGeodes(actions, timeLeft - timeToWait, newResources, newResourcesPerMinute, statesVisited, depth + 1);
        bestGeodes = Math.max(bestGeodes, geodes);
    }
    return bestGeodes;
}

async function solve(filename: string) {
    const filepath = path.join(__dirname, filename);
    const inputText = (await fs.readFile(filepath, 'utf-8')).trimEnd();

    console.log(`Solving for ${filename}...`);

    const lines = inputText.split('\n');
    let qualityNumberSum = 0;
    for (const [i, line] of lines.entries()) {
        const actions = parseCostsToActions(line);

        process.stdout.write(`Blueprint ${i + 1}: `);

        // console.log(actions);

        // Start with no resources.
        const resources = new Map(ZERO_RESOURCE_MAP);
        // But we do start with one ore-producing robot.
        const resourcesPerMinute = new Map(ZERO_RESOURCE_MAP);
        resourcesPerMinute.set('ore', 1);

        const geodes = maximiseGeodes(actions, 24, resources, resourcesPerMinute, new Map());
        const qualityNumber = geodes * (i + 1);

        qualityNumberSum += qualityNumber;

        console.log(`${geodes} geodes. Quality number: ${qualityNumber} (sum = ${qualityNumberSum}))`);
    }

    console.log(`Part 1: ${qualityNumberSum}`);
}

Promise.resolve()
.then(() => solve('demo.txt'))
// .then(() => solve('input.txt'))

File: 19/robots.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

var resources = [4]string{"ore", "clay", "obsidian", "geode"}

var resourceIndices = map[string]int{
	"ore":      0,
	"clay":     1,
	"obsidian": 2,
	"geode":    3,
}

// A blueprint is the cost of each robot type.
type Blueprint [4][4]int

type SearchState struct {
	TimeLeft           int
	Resources          [4]int
	ResourcesPerMinute [4]int
}

// Parses the cost of each robot type from a line of the input using regex.
//
// Example line:
// `Blueprint 1: Each ore robot costs 4 ore. Each clay robot costs 4 ore. Each obsidian robot costs 4 ore and 17 clay. Each geode robot costs 4 ore and 20 obsidian.`
func parseBlueprint(line string) Blueprint {
	// Regex to match one sentence describing one robot.
	// Example: `Each ore robot costs 4 ore.`
	robotRegex := regexp.MustCompile(`Each (\w+) robot costs (.+?)\.`)
	// Find all the sentences describing the robots.
	robotMatches := robotRegex.FindAllStringSubmatch(line, -1)
	// Create a new blueprint
	bp := Blueprint{}
	// For each robot, parse the cost
	for _, robotMatch := range robotMatches {
		robotType := robotMatch[1]
		robotTypeIndex := resourceIndices[robotType]
		// Parse the cost of the robot
		cost := parseCost(robotMatch[2])
		// Set the cost in the blueprint
		bp[robotTypeIndex] = cost
	}
	return bp
}

// Parses a cost string into an array of ints.
//
// Example cost string: `4 ore and 17 clay`
func parseCost(cost string) [4]int {
	// Regex to match one cost component.
	// Example: `4 ore`
	costRegex := regexp.MustCompile(`(\d+) (\w+)`)
	// Find all the cost components
	costMatches := costRegex.FindAllStringSubmatch(cost, -1)
	// Create a new cost array
	c := [4]int{}
	// For each cost component, set the cost in the array
	for _, costMatch := range costMatches {
		// Parse the cost
		value, err := strconv.Atoi(costMatch[1])
		if err != nil {
			panic(err)
		}
		// Set the cost in the array
		resource := costMatch[2]
		resourceIndex := resourceIndices[resource]
		c[resourceIndex] = value
	}
	return c
}

func getSubStates(state SearchState, bp Blueprint) []SearchState {
	subStates := make([]SearchState, 0)
	// For each type of robot, try waiting until we have enough resources and then building it.
outer:
	// r = robot
	for r := 0; r < 4; r++ {
		timeToWait := 0
		// For each resource, figure out how long we'd need to wait until we have enough.
		// If there's no way we can wait long enough, give up.

		// rr = robot resource
		for rr := 0; rr < 4; rr++ {
			amountLeft := bp[r][rr] - state.Resources[rr]
			if amountLeft <= 0 {
				continue
			}
			if amountLeft > 0 && state.ResourcesPerMinute[rr] == 0 {
				// We can't wait long enough to get enough of this resource.
				continue outer
			}
			// Ceiling division
			timeToWaitForResource := (amountLeft + state.ResourcesPerMinute[rr] - 1) / state.ResourcesPerMinute[rr]
			if timeToWaitForResource > timeToWait {
				timeToWait = timeToWaitForResource
			}
		}
		// We have to wait 1 extra minute to build the robot.
		timeToWait += 1

		// If we don't have enough time to wait, don't try
		if timeToWait > state.TimeLeft {
			continue
		}

		// Create a new state with the robot built
		newState := SearchState{
			TimeLeft:           state.TimeLeft - timeToWait,
			Resources:          state.Resources,
			ResourcesPerMinute: state.ResourcesPerMinute,
		}
		// For each resource, add the amount we'd get from waiting
		for rr := 0; rr < 4; rr++ {
			newState.Resources[rr] += timeToWait * state.ResourcesPerMinute[rr]
		}
		// Subtract the cost of the robot
		for rr := 0; rr < 4; rr++ {
			newState.Resources[rr] -= bp[r][rr]
		}
		// Add the resources we get from the robot
		newState.ResourcesPerMinute[r] += 1
		subStates = append(subStates, newState)
	}
	return subStates
}

// The score of this state if we did nothing else.
func scoreState(state SearchState) int {
	// We just care about how many geodes we'd have at the end.
	return state.Resources[3] + state.TimeLeft*state.ResourcesPerMinute[3]
}

// An upper bound of how good this state could possibly be.
func upperBound(state SearchState, bp Blueprint) int {
	// Perhaps not to far off the actual value, but lets assume that from here we can build a geode robot every minute.
	// The amount of geodes we'd have grows with a triangle pattern.
	// Given it takes 1 minute before the robot produces anything, we use state.TimeLeft - 1 in the equation.
	return scoreState(state) + state.TimeLeft*(state.TimeLeft-1)/2
}

func findMaxGeodes(bp Blueprint, time int) int {
	mostGeodes := 0
	// The initial state is we have no resources, but 1 ore robot.
	initialState := SearchState{
		TimeLeft:           time,
		Resources:          [4]int{},
		ResourcesPerMinute: [4]int{1, 0, 0, 0},
	}

	// For debugging
	searched := 0
	skipped := 0

	// Do a depth-first search of the search space.
	// Depth first is good because we can prune branches early.
	// We use a stack to do the search.
	stack := []SearchState{initialState}
	for len(stack) > 0 {
		// Pop the top state off the stack
		state := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		// If this state is better than the best we've seen, update the best
		if score := scoreState(state); score > mostGeodes {
			fmt.Println("New best:", score, "searched:", searched, "skipped:", skipped)
			mostGeodes = score
		}

		// If this state could not be better than the best we've seen, don't bother exploring it.
		if upperBound(state, bp) <= mostGeodes {
			skipped++
			continue
		}

		// Get all the states we can get to from this state
		subStates := getSubStates(state, bp)
		// Push them onto the stack
		stack = append(stack, subStates...)
		searched++
	}

	fmt.Println("searched:", searched, "skipped:", skipped)

	return mostGeodes
}

func solvePt1(filename string) int {
	// Read the file, line by line
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	qualitySum := 0

	scanner := bufio.NewScanner(file)
	for i := 0; scanner.Scan(); i++ {
		// Parse the blueprint
		bp := parseBlueprint(scanner.Text())
		// Find the maximum number of geodes we can get
		mostGeodes := findMaxGeodes(bp, 24)
		quality := mostGeodes * (i + 1)
		qualitySum += quality
		fmt.Println("Geodes:", mostGeodes, "Quality:", quality, "Sum:", qualitySum)
	}

	return qualitySum
}

func solvePt2(filename string) int {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	mostGeodesMultiplied := 1

	scanner := bufio.NewScanner(file)
	// Just read the first 3 lines.
	for i := 0; i < 3 && scanner.Scan(); i++ {
		// Parse the blueprint
		bp := parseBlueprint(scanner.Text())
		// Find the maximum number of geodes we can get
		mostGeodes := findMaxGeodes(bp, 32)
		mostGeodesMultiplied *= mostGeodes
		fmt.Println("Geodes:", mostGeodes)
	}

	return mostGeodesMultiplied
}

func main() {
	filename := "19/input.txt"
	pt1 := solvePt1(filename)
	fmt.Println("Pt1:", pt1)

	pt2 := solvePt2(filename)
	fmt.Println("Pt2:", pt2)
}
